### DISCUSSION TITLE
OOM issue with Next.js on Kubernetes

### DISCUSSION THREAD

**Discussion by erkand-imeri** [vercel/next.js#46873] [Help]
### Summary

Hello all, i am facing OOM issue with Next.js on Kubernetes.

Pod memory keeps growing but it doesn't revert back to previous memory, it keeps growing linearly until it reaches a certain limit and Kubernetes restarts the pods.

Initially i thought it might be something to do with the code, garbage collector not being able to free up the memory but it doesn't look so, because i deployed another app, plain Next.js which renders just a random text and has minimal dependencies and behaves the same.

Now, i am Dockerizing my application, my apps are in a monorepo locally managed by turborepo, and the dockerized application i am hosting them/deploying in a Kubernetes cluster.

```js
ARG NODE=node:16.15.0-alpine

FROM ${NODE} as deps
RUN yarn set version 1.22.19

WORKDIR /app
COPY apps/plain-nextjs/ ./
RUN yarn install

FROM ${NODE} as builder
ENV NODE_ENV production 

WORKDIR /app
COPY --from=deps /app/ ./
RUN yarn build

FROM ${NODE} as production
WORKDIR /app
ENV NODE_ENV production 


RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/next.config.js ./
COPY --from=builder /app/package.json ./package.json
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3005

ENV PORT 3005

CMD node --max-old-space-size=360 server.js

```

The pod memory limit in K8 helm charts is defined as 512MB, so in Node.js/Next.js i assign 75% of that, based on recommendations online.

I am also using path based routing from Kubernetes, i use 1 domain/URL for several related mini-web-apps which are separated by routes, istio path based routing.

In next.config.js i defined the following:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  webpack: (config, options) => {
    config.module.rules.push({
      test: /\.svg$/,
      use: ["@svgr/webpack", "url-loader"],
    });
    return config;
  },
  reactStrictMode: false,
  output: "standalone",
  swcMinify: true,
  compress: true,
  basePath: "/plain-nextjs",
};

module.exports = nextConfig;

```

Say that we use a domain https://mydomain.internal.io, for this particular app we setup in istio virtual-service to actually be redirected when users browse:

```js
https://mydomain.internal.io/plain-nextjs
```

Point is, i thought initially the memory leak happens because of some circular imports, or some garbage collection issue.

I cannot make sense and point finger why this is happening. Anyone might have any idea?

In addition, this is the package.json

```js
{
  "name": "plain-nextjs",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev -p 3005",
    "build": "next build",
    "start": "next start -p 3005",
    "test": "jest --forceExit --detectOpenHandles",
    "lint": "next lint"
  },
  "dependencies": {
    "@next/font": "13.1.5",
    "next": "13.1.5",
    "prop-types": "^15.8.1",
    "react": "18.2.0",
    "react-dom": "18.2.0"
  },
  "devDependencies": {
    "@svgr/webpack": "^6.5.1",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "babel-plugin-inline-react-svg": "^2.0.1",
    "eslint": "8.32.0",
    "eslint-config-next": "13.1.5",
    "http-status-codes": "^2.2.0",
    "jest": "^28.1.3",
    "jest-environment-jsdom": "^29.4.0",
    "styled-components": "^5.3.6",
    "url-loader": "^4.1.1"
  }
}
```

**Comment by erkand-imeri** [comment]
Hello, so i am posting the github repo of plain-nextjs app where apparently i am facing OOM issue:

https://github.com/erkand-imeri/plain-nextjs-k8

And a screenshot from my staging environment, grafana visualization/chart of the memory consumption, the staging load is not big but the pattern is visible, pod memory keeps increasing and doesn't resort to its initial state. 

![Screenshot from 2023-03-09 11-21-11](https://user-images.githubusercontent.com/14806022/224006578-b37855fe-a684-4189-a7d5-89f74228f1bd.png)

**Reply by icyJoseph** [reply]
Nice, its is often the case that one needs a reproduction demo.

- Have we discarded that this is not a pod problem?
  - Take this question for example, https://stackoverflow.com/questions/73293227/pods-memory-is-not-released-after-high-load-is-gone, someone even says that a Rust app is experiencing the same issue
- https://github.com/aws/amazon-ecs-agent/issues/2865


Could you try to implement this endpoint: https://github.com/vercel/next.js/discussions/26801#discussioncomment-1091592

And get the heap dumps?

**Comment by erkand-imeri** [comment]
Hello, @icyJoseph, i think i tried that one:

```js
experimental: {
 isrMemoryCacheSize: 0 
}
```

and made no difference for me.

//EDIT: You meant the headdump, i will try that one.

One thing i stumbled is the OOM could happen from memory leak but also from lazy garbage collection, somehow i think the second option is more viable in my case, there is no indication Next.js has a memory leak, i am just rendering a text.

In my Dockerfile i am defining this:

```js
CMD node --max-old-space-size=360 server.js

```

And in my helm charts

```js
resources:
  requests:
    cpu: 50m
    memory: 128Mi
  limits:
    cpu: 300m
    memory: 512Mi
```

So the limit in Kubernetes for the memory is defined with 512MB. Question is am i defining it correctly, is there any correlation of resources.limits and --max-old-space-size node flag? Could it be that Node.js garbage collector is just ignoring the Kubernetes limit and lazy garbage collecting but Kubernetes restarts the pod when memory limit is reached?

**Reply by icyJoseph** [reply]
I didn't mean to try that solution, but to implement the endpoint he shared on the discussion, where he's able to collect the heap dump from the runtime.

**Reply by icyJoseph** [reply]
My bad, I shared a link to the answer and not the code... https://github.com/vercel/next.js/discussions/26801#discussioncomment-1094215 there

**Reply by icyJoseph** [reply]
Interesting take, so there would be two limits conflicting each other... I'll investigate!

**Comment by erkand-imeri** [comment]
Based on this description from cloud.google

```js
requests.memory is the maximum combined Memory requests for all the containers in the Namespace. In the above example, you can have 50 containers with 2MiB requests, five containers with 20MiB CPU requests, or even a single container with a 100MiB request. As long as the total requested Memory in the Namespace is less than 100MiB!

https://cloud.google.com/blog/products/containers-kubernetes/kubernetes-best-practices-resource-requests-and-limits
```

Does it mean that when we declare in Kubernetes the resource limit say to 512MB that should reflect how many containers/pods we declare.

For instance minimum pods i declare in my app are 2 pods, 1 pod is mirrored with 1 container.

And 1 container in Dockerfile takes:

```js
CMD node --max-old-space-size=360 server.js
```

I am wondering if i should split the Kubernetes memory limit in Docker image context by 2.

So instead of thinking i have 512MB i should consider it 256MB per pod. And in the pod declare 200MB -max-old-space-size.

I will test this and see how it behaves. But, my assumption might be wrong. Not sure about this.

**Comment by erkand-imeri** [comment]
Hey all, so the main issue with this was not memory leak from the app but rather likely lazy garbage collection.

So, when in Kubernetes you define resource limits:

```js
resources:
  requests:
    cpu: 50m
    memory: 128Mi
  limits:
    cpu: 300m
    memory: 512Mi
```

This means that your app per pod has 512Mi limit, but then if you are using istio/sidecar, istio proxy hence service mesh, for each app pod one istio-proxy pod gets created. In my case one istio-proxy pod was taking 150Mi of memory.

You have to take in consideration that your app pod will not take the whole memory defined in K8 resource limit object, it is usually recommended per app pod to define the --max-old-space-size 80% of what you define in K8, i would take that 80% and substract the istio-pod memory and then assign it to app pod memory in Dockerfile.

```js
CMD node --max-old-space-size=200 --optimize_for_size --gc_interval=100 server.js
```

There is not a generic rule for this, you should play around with --max-old-space-size and --gc_interval to see what works better for your service.

Furthemore, i highly recommend when using istio-proxy to define Sidecar objects to scope down the namespaces your service/deployment uses the mesh service with.

```js
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: default
  namespace: your-namespace
spec:
  egress:
  - hosts:
    - "your-namespace/*"
```

This will reduce the memory size of istio-proxy and will make the app pod memory to behave more stable. Or you can disable sidecar.istio.io/injection if your app/service/deployment don't uses istio service mesh at all.

In helm files, in deployment.yaml

```js
template:
    metadata:
      labels:
        sidecar.istio.io/inject: "false"
```

This will disable istio-proxy pods from being created, your app memory will reflect 100% what you defined in Kubernetes resources. It is usually recommended to use istio-proxy because it acts like load balancer as well. But, it's an option, depending on specific service needs.
